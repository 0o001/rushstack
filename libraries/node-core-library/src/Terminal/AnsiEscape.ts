// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Options for {@link AnsiEscape.formatForTests}.
 * @public
 */
export interface IAnsiEscapeConvertForTestsOptions {
  /**
   * If true then `\n` will be replaced by `[n]`, and `\r` will be replaced by `[r]`.
   */
  encodeNewlines?: boolean;
}

/**
 * Operations for encoding and decoding text strings containing
 * {@link https://en.wikipedia.org/wiki/ANSI_escape_code | ANSI escape codes}.
 * The most commonly used escape codes set the foreground/background color for
 * console output.
 * @public
 */
export class AnsiEscape {
  // For now, we only care about the Control Sequence Introducer (CSI) commands which always start with "[".
  // eslint-disable-next-line no-control-regex
  private static _csiRegExp: RegExp = /\x1b\[([\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])/gu;

  // Text coloring is performed using Select Graphic Rendition (SGR) codes, which come after the
  // CSI introducer "ESC [".  The SGR sequence is a number followed by "m".
  private static _sgrRegExp: RegExp = /([0-9]+)m/u;

  /**
   * Replaces ANSI escape codes with human-readable tokens.  This is useful for unit tests
   * that compare text strings in test assertions or snapshot files.
   */
  public static formatForTests(text: string, options?: IAnsiEscapeConvertForTestsOptions): string {
    if (!options) {
      options = {};
    }

    let result: string = text.replace(AnsiEscape._csiRegExp, (capture: string, csiCode: string) => {
      const match: RegExpMatchArray | null = csiCode.match(AnsiEscape._sgrRegExp);

      if (match) {
        const sgiCode: number = parseInt(match[1]);
        const colorCode: string | undefined = AnsiEscape._x(sgiCode);
        if (colorCode) {
          // Example: "[black-bg]"
          return `[${colorCode}]`;
        }
      }

      // Example: "[31m]"
      return `[${csiCode}]`;
    });

    if (options.encodeNewlines) {
      result = result.replace(/\n/g, '[n]').replace(/\r/g, `[r]`);
    }
    return result;
  }

  private static _x(n: number): string | undefined {
    switch (n) {
      case 30:
        return 'black';
      case 31:
        return 'red';
      case 32:
        return 'green';
      case 33:
        return 'yellow';
      case 34:
        return 'blue';
      case 35:
        return 'magenta';
      case 36:
        return 'cyan';
      case 37:
        return 'white';
      case 90:
        return 'gray';
      case 39:
        return 'default';
      case 40:
        return 'black-bg';
      case 41:
        return 'red-bg';
      case 42:
        return 'green-bg';
      case 43:
        return 'yellow-bg';
      case 44:
        return 'blue-bg';
      case 45:
        return 'magenta-bg';
      case 46:
        return 'cyan-bg';
      case 47:
        return 'white-bg';
      case 100:
        return 'gray-bg';
      case 49:
        return 'default-bg';
      case 1:
        return 'bold';
      case 21:
        return 'bold-off';
      case 2:
        return 'dim';
      case 22:
        return 'normal';
      case 4:
        return 'underline';
      case 24:
        return 'underline-off';
      case 5:
        return 'blink';
      case 25:
        return 'blink-off';
      case 7:
        return 'invert';
      case 27:
        return 'invert-off';
      case 8:
        return 'hidden';
      case 28:
        return 'hidden-off';
      default:
        return undefined;
    }
  }
}
