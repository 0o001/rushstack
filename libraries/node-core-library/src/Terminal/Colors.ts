// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * @beta
 */
export interface IColorableSequence {
  text: string;
  foregroundColor?: ColorValue;
  backgroundColor?: ColorValue;
}

/**
 * @beta
 */
export enum ColorValue {
  Black,
  Red,
  Green,
  Yellow,
  Blue,
  Magenta,
  Cyan,
  White,
  Gray
}

/**
 * @beta
 */
export class Colors {
  public static black(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Black
    };
  }

  public static red(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Red
    };
  }

  public static green(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Green
    };
  }

  public static yellow(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Yellow
    };
  }

  public static blue(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Blue
    };
  }

  public static magenta(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Magenta
    };
  }

  public static cyan(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Cyan
    };
  }

  public static white(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.White
    };
  }

  public static gray(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      foregroundColor: ColorValue.Gray
    };
  }

  public static blackBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Black
    };
  }

  public static redBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Red
    };
  }

  public static greenBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Green
    };
  }

  public static yellowBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Yellow
    };
  }

  public static blueBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Blue
    };
  }

  public static magentaBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Magenta
    };
  }

  public static cyanBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Cyan
    };
  }

  public static whiteBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.White
    };
  }

  public static grayBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...(typeof text === 'string' ? { text } : text),
      backgroundColor: ColorValue.Gray
    };
  }
}