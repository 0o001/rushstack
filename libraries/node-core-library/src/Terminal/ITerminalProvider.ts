/**
 * @beta
 */
export enum Severity {
  log,
  warning,
  error,
  verbose
}

/**
 * @beta
 */
export interface ITerminalProvider {
  supportsColor: boolean;
  write(data: string, severity: Severity): void;
}
