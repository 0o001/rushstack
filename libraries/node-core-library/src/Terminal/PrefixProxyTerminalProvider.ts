// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';

/**
 * Options for {@link PrefixProxyTerminalProvider}.
 *
 * @beta
 */
export interface IPrefixProxyTerminalProviderOptions {
  /**
   * The {@link ITerminalProvider} that will be wrapped.
   */
  terminalProvider: ITerminalProvider;

  /**
   * The prefix that should be added to each line of output.
   */
  prefix: string;
}

function escapeRegExp(literal: string): string {
  // Perform a cheap escape of the literal string so that it can be used in a RegExp
  return literal.replace(/[^A-Za-z0-9_]/g, '\\$&');
}

/**
 * Wraps an existing {@link ITerminalProvider} that prefixes each line of output with a specified
 * prefix string.
 *
 * @beta
 */
export class PrefixProxyTerminalProvider implements ITerminalProvider {
  private _parentTerminalProvider: ITerminalProvider;
  private _prefix: string;
  private _currentPrefix: string;
  private _newlineRegex: RegExp;

  public constructor(options: IPrefixProxyTerminalProviderOptions) {
    const { terminalProvider, prefix } = options;
    this._parentTerminalProvider = terminalProvider;
    this._prefix = prefix;
    this._currentPrefix = prefix;
    // eslint-disable-next-line @rushstack/security/no-unsafe-regexp
    this._newlineRegex = new RegExp(`${escapeRegExp(terminalProvider.eolCharacter)}|\\n`, 'g');
  }

  /** @override */
  public get supportsColor(): boolean {
    return this._parentTerminalProvider.supportsColor;
  }

  /** @override */
  public get eolCharacter(): string {
    return this._parentTerminalProvider.eolCharacter;
  }

  /** @override */
  public write(data: string, severity: TerminalProviderSeverity): void {
    // We need to track newlines to ensure that the prefix is added to each line
    let currentIndex: number = 0;
    // eslint-disable-next-line @rushstack/no-new-null
    let newlineMatch: RegExpExecArray | null;
    while ((newlineMatch = this._newlineRegex.exec(data))) {
      // Extract the line, add the prefix, and write it out with the newline
      const newlineIndex: number = newlineMatch.index;
      const newIndex: number = newlineIndex + newlineMatch[0].length;
      const dataToWrite: string = `${this._currentPrefix}${data.substring(currentIndex, newIndex)}`;
      this._parentTerminalProvider.write(dataToWrite, severity);
      // Update the currentIndex to start the search from the char after the newline
      currentIndex = newIndex;
      this._currentPrefix = this._prefix;
    }

    // The remaining data is not postfixed by a newline, so write out the data and set _isNewline to false
    const remainingData: string = data.substring(currentIndex);
    if (remainingData.length) {
      this._parentTerminalProvider.write(`${this._currentPrefix}${remainingData}`, severity);
      this._currentPrefix = '';
    }
  }
}
