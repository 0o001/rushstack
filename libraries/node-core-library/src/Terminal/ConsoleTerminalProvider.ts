// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { enabled as supportsColor } from 'colors/safe';

import { ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';

export interface IConsoleTerminalProviderOptions {
  verboseEnabled: boolean;
}

/**
 * @beta
 */
export class ConsoleTerminalProvider implements ITerminalProvider {
  public verboseEnabled: boolean = false;

  public constructor(options: Partial<IConsoleTerminalProviderOptions> = {}) {
    this.verboseEnabled = !!options.verboseEnabled;
  }

  public write(data: string, severity: TerminalProviderSeverity): void {
    switch (severity) {
      case TerminalProviderSeverity.warning:
      case TerminalProviderSeverity.error: {
        process.stderr.write(data);
        break;
      }

      case TerminalProviderSeverity.verbose: {
        if (this.verboseEnabled) {
          process.stdout.write(data);
        }
        break;
      }

      case TerminalProviderSeverity.log:
      default: {
        process.stdout.write(data);
        break;
      }
    }
  }

  public get supportsColor(): boolean {
    return supportsColor;
  }
}
