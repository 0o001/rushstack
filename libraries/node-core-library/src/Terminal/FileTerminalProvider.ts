import { ITerminalProvider } from './ITerminalProvider';
import { FileSystem } from '../FileSystem';

/**
 * @beta
 */
export class FileTerminalProvider implements ITerminalProvider {
  private _filePath: string;

  public constructor(filePath: string) {
    this._filePath = filePath;
  }

  public write(data: string): void {
    try {
      FileSystem.appendToFile(this._filePath, data);
    } catch (e) {
      // Ignore
    }
  }

  public get width(): number | undefined {
    return Number.MAX_VALUE;
  }

  public get supportsColor(): boolean {
    return false;
  }
}
