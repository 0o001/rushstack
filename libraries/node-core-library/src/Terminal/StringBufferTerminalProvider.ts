// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminalProvider, Severity } from './ITerminalProvider';

/**
 * @beta
 */
export class StringBufferTerminalProvider implements ITerminalProvider {
  private _standardBuffer: string[] = [];
  private _warningBuffer: string[] = [];
  private _errorBuffer: string[] = [];

  private _supportsColor: boolean;

  public constructor(supportsColor: boolean = false) {
    this._supportsColor = supportsColor;
  }

  public write(data: string, severity: Severity): void {
    switch (severity) {
      case Severity.warn: {
        this._warningBuffer.push(data);
        break;
      }

      case Severity.error: {
        this._errorBuffer.push(data);
        break;
      }

      case Severity.log:
      default: {
        this._standardBuffer.push(data);
        break;
      }
    }
  }

  public get width(): number | undefined {
    return process.stdout.columns;
  }

  public get supportsColor(): boolean {
    return this._supportsColor;
  }

  public clear(): void {
    this._standardBuffer = [];
    this._warningBuffer = [];
    this._errorBuffer = [];
  }

  public getOutput(): string {
    const combinedBuffer: string = this._standardBuffer.join('');
    this._standardBuffer = [combinedBuffer];
    return combinedBuffer;
  }

  public getErrorOutput(): string {
    const combinedBuffer: string = this._errorBuffer.join('');
    this._standardBuffer = [combinedBuffer];
    return combinedBuffer;
  }

  public getWarningOutput(): string {
    const combinedBuffer: string = this._warningBuffer.join('');
    this._standardBuffer = [combinedBuffer];
    return combinedBuffer;
  }
}
