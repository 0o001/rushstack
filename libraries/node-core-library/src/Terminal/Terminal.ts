import { ITerminalProvider } from "./ITerminalProvider";

export class Terminal {
  private _provider;

  public constructor(provider: ITerminalProvider) {
    this._provider = provider;
  }

  public write(message: string) {}
  public writeWarning(message: string) {}
  public writeError(message: string) {}
  public writeVerbose(message: string) {}
}
