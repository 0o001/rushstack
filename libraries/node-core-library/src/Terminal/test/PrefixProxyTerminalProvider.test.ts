// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal } from '../Terminal';
import { StringBufferTerminalProvider } from '../StringBufferTerminalProvider';
import { PrefixProxyTerminalProvider } from '../PrefixProxyTerminalProvider';

let terminal: Terminal;
let provider: StringBufferTerminalProvider;

function verifyProvider(): void {
  expect({
    log: provider.getOutput(),
    warning: provider.getWarningOutput(),
    error: provider.getErrorOutput(),
    verbose: provider.getVerbose(),
    debug: provider.getDebugOutput()
  }).toMatchSnapshot();
}

describe(PrefixProxyTerminalProvider.name, () => {
  beforeEach(() => {
    provider = new StringBufferTerminalProvider(true);
    const prefixProvider: PrefixProxyTerminalProvider = new PrefixProxyTerminalProvider({
      terminalProvider: provider,
      prefix: '[prefix] '
    });
    terminal = new Terminal(prefixProvider);
  });

  describe('write', () => {
    test('writes a message', () => {
      terminal.write('test message');
      verifyProvider();
    });

    test('writes a message with newlines', () => {
      terminal.write('message 1\nmessage 2\nmessage 3');
      verifyProvider();
    });

    test('writes a message with provider newlines', () => {
      terminal.write(`message 1${provider.eolCharacter}message 2${provider.eolCharacter}message 3`);
      verifyProvider();
    });

    test('writes messages without newlines', () => {
      terminal.write('message 1');
      terminal.write('message 2');
      terminal.write('message 3');
      verifyProvider();
    });

    test('writes a mix of messages with and without newlines', () => {
      terminal.write('message 1');
      terminal.write('message 2\nmessage 3\n');
      terminal.write('message 4');
      terminal.write('message 5\nmessage 6');
      verifyProvider();
    });
  });

  describe('writeLine', () => {
    test('writes a message line', () => {
      terminal.writeLine('test message');
      verifyProvider();
    });

    test('writes a message line with newlines', () => {
      terminal.writeLine('message 1\nmessage 2\nmessage 3');
      verifyProvider();
    });

    test('writes a message line with provider newlines', () => {
      terminal.writeLine(`message 1${provider.eolCharacter}message 2${provider.eolCharacter}message 3`);
      verifyProvider();
    });

    test('writes a mix of message lines with and without newlines', () => {
      terminal.writeLine('message 1');
      terminal.writeLine('message 2\nmessage 3\n');
      terminal.writeLine('message 4');
      terminal.writeLine('message 5\nmessage 6');
      verifyProvider();
    });
  });
});
