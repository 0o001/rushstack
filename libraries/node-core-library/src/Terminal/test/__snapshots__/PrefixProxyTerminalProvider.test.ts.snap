// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrefixProxyTerminalProvider write writes a message 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] test message",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider write writes a message with newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1[n][prefix] message 2[n][prefix] message 3",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider write writes a message with provider newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1[n][prefix] message 2[n][prefix] message 3",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider write writes a mix of messages with and without newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1message 2[n][prefix] message 3[n][prefix] message 4message 5[n][prefix] message 6",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider write writes messages without newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1message 2message 3",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider writeLine writes a message line 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] test message[n]",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider writeLine writes a message line with newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1[n][prefix] message 2[n][prefix] message 3[n]",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider writeLine writes a message line with provider newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1[n][prefix] message 2[n][prefix] message 3[n]",
  "verbose": "",
  "warning": "",
}
`;

exports[`PrefixProxyTerminalProvider writeLine writes a mix of message lines with and without newlines 1`] = `
Object {
  "debug": "",
  "error": "",
  "log": "[prefix] message 1[n][prefix] message 2[n][prefix] message 3[n][prefix] [n][prefix] message 4[n][prefix] message 5[n][prefix] message 6[n]",
  "verbose": "",
  "warning": "",
}
`;
